Introduce OIDC Database Token State Manager extension

michalvavrik opened the pull request:
closes: #13239

New extension uses the Reactive SQL clients to store and retrieve tokens in a database, instead of storing them in a session cookie (which is default behavior). Currently Hibernate ORM and Hibernate Reactive doesn't work together, which means I could not leverage Hibernate Reactive, for the most applications are likely to use the Hibernate ORM extension directly. On the other hand, basing this on JDBC driver and block during authentication seems like a very bad design, because it would lead to thread switching. I don't think we should provide something so wrong (blocking auth), users can always write it themselves if necessary. As far as suggestion to use Panache goes, I also think it's not suitable (1. Hibernate Reactive vs ORM, 2. multitenancy is not supported).

michalvavrik commented:
Wow, these failures by name seems very much related, though I only extracted 2 methods from OIDC module. I'm going to check today how that could result in this.

sberyozkin commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcUtils.java:
```diff
@@ -532,6 +532,24 @@ public static String decryptString(String jweString, Key key) throws Exception {
         return decryptString(jweString, key, KeyEncryptionAlgorithm.A256GCMKW);
     }
 
+    public static String decryptToken(String token, RoutingContext context) {
```
@michalvavrik  I'm not sure the db state manager needs to worry about encrypting/decrypting - it is only relevant for the situation, where we have tokens in the cookie, and since the cookie, with this option, will just have a DB pointer, we don't really need to encrypt/decrypt then before saving in the database or encrypting decrypting this cookie value.

michalvavrik replied:
@sberyozkin hey Sergey, I think there is possible compromise - I'll change PK to database generated UUID. I'll do it tonight. This way, no encryption is needed, while it's still safe.

michalvavrik replied:
@sberyozkin please hold with the Quarkus-dev e-mail till I fix that tonight. thank you!

michalvavrik replied:
pity I didn't realize this before :-)

sberyozkin replied:
>in case they have direct access to database

Then it is already all has gone wrong, we can encrypt as much as we want but they will just access the database.

>then next cookie value for other user can't be guessed!

Sure, I thought it was a random value, each new record key, as opposed to some predictable increment. But even if they guess the next value, what is the risk  ?

How would it be different, conceptually, from the traditional stateful applications where the cookie is essentially a pointer to some server side state ?

That said, indeed, lets keep it as you have thought about it for now, and discuss a long the way

Thanks

michalvavrik replied:
@sberyozkin I'm going to drop encryption and use UUID instead

michalvavrik replied:
> Then it is already all has gone wrong, we can encrypt as much as we want but they will just access the database.

I had in my mind DBA that can access DB, but not app secret etc. Anyway, let's go with DB-generated UUID.

michalvavrik replied:
> Sure, I thought it was a random value, each new record key, as opposed to some predictable increment. But even if they guess the next value, what is the risk ?

I would need to spend more time on this, but I was afraid that someone will be authenticated, but will have access to someone else tokens pointer. Maybe that's rubbish, it is better to make it random.

sberyozkin commented:
Hey @michalvavrik Thanks for preparing this new feature :+1: , a few tests are failing :-), but I think they'll get green as I believe you don't need to change anything in the default token manager, there is nothing to secure in the cookie value returned from the DB one - tokens are secured in the DB.

I'll prepare an email to the quarkus-dev with some justifications for the new feature

Thanks

emmanuelbernard commented:
Is it a 30/70% PR review state or aimed ot be merged? If the latter,m I'd love ot see the doc changes and how we plan to expose that to users.

michalvavrik commented:
> Is it a 30/70% PR review state or aimed ot be merged? If the latter,m I'd love ot see the doc changes and how we plan to expose that to users.

I'll let @sberyozkin answer, I suppose Sergey will review when he has time. Only thing I know about now is that I have one unused property in recorder that I plan to remove when someone requires any changes at all.

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerRunTimeConfig.java:
```diff
@@ -0,0 +1,33 @@
+     * You can disable this feature when you want only one of your service instances to delete expired tokens.
+     */
+    @WithDefault("true")
+    boolean deleteExpired();
```
I'm not sure that this property is required. If the cookie representing this DB data itself has expired the browser will remove it and it won't reach Quarkus and `TokenStateManager#remove` will not be even called. I.e, if this property is `false`, DB will slowly keep growing

michalvavrik replied:
> I'm not sure that this property is required. If the cookie representing this DB data itself has expired the browser will remove it and it won't reach Quarkus and `TokenStateManager#remove` will not be even called. I.e, if this property is `false`, DB will slowly keep growing

Consider following situations:

- in OpenShift you are most likely to run your service in more than one track. E.g. you scale to 2-3, whatever works for you. but it doesn't make sense for every pod to try to delete expired tokens, there is no additional value, hence you only enable it on one instance (it can be automated, you will only set this env var on one of pods via deployment config)
- you definitely won't to delete tokens that have expired and browser removed them, but would stick in our database in cases when TokenStateManager#deleteToken wasn't called. it could be potential point of attack if database keep growing if the attacker was authenticated, he would just drop the old token and authenticated again, and again, until DoS succeed

That's way it is there.

michalvavrik replied:
BTW I can imagine other, though less likely scenarios when you want to stop delete. You can have service with cluster schedulers and do periodic stuff only in there, so users will delete expired tokens themselves. Or someone may prefer trigger (it wouldn't be very performant though).

I suppose your concern is rather about keeping as few configuration options as possible? I don't have strong opinion there, but for arguments I provided, I vote to keep this flag.

michalvavrik replied:
> My point is, if this property is ever set to false then how the expired tokens will be removed, given that we know that if the only link to the data in the form of the cookie has been deleted by the browser, ex, when the user cleared the cache or the cookie has expired ?

Yeah, I'm having trouble to find words we would both understand. Hope you won't get offended if I try to simplify it:

- create quarkus project hello-app
- deploy it to openshift -> hello-app-svc
- scale to 2 pods
- after 8 h, both pods are deleting expired tokens; I'm not worried about race, but it doesn't make sense. 

now instead, I create 2 deployment configs, both will scale to one (2x1=2):
-  with env var that disables deletion
- without env var that will delete tokens, this deployment config will never scale to zero because you always want to keep such service up

I think that is the very reason why cluster schedulers exists, so that you don't execute in all app instances, but just one of them. And so, if someone has already app let say with Quartz job, he can also disable deletion and instead run delete statement himself. (I didn't document this last paragraph, for I think it's unlikely, it would just add noise to docs).

sberyozkin replied:
@michalvavrik 

>after 8 h, both pods are deleting expired tokens

I wonder if if it is where the problem is. I.e, my understanding we don't introduce this feature as Kubernetes native feature - for such multi-pods set ups, well, let users have it in the encrypted cookies, the problem is solved. We offer it for simple cases, no high throughput flows.

Or if the users insist on using it in Kubernetes, then like you said, both pods will send delete requests to the DB - is there a problem there apart it is suboptimal ? At least it is guaranteed to happen, as opposed to depending on some human intervention to prepare the scheduler specifically to manage the deletion of the token records. 

I'd not mind, at least for now, dropping this property and reintroduce it, now that I know why you considered it, when someone actually tries to use the feature in the cloud in a multi-pod setup, and considers it being suboptimal.

sberyozkin replied:
If you prefer to keep then IMHO to make it very clear in the docs that users should not disable it in most cases, only consider doing in a set up you have described with the scheduler being used instead. 
That should be ok then I guess, it will prevent users from trying do some optimizations which will lead to the leaks.
I'll let you decide, I'll support

michalvavrik replied:
> Or if the users insist on using it in Kubernetes, then like you said, both pods will send delete requests to the DB - is there a problem there apart it is suboptimal ? At least it is guaranteed to happen, as opposed to depending on some human intervention to prepare the scheduler specifically to manage the deletion of the token records.

No, it's not an issue, really. You convinced me, we can drop it. The documentation part, also makes sense, it would need to be better documented. **I am going to drop the flag.**

> for such multi-pods set ups, well, let users have it in the encrypted cookies, the problem is solved. We offer it for simple cases, no high throughput flows.

I never mentioned it before, for you keep mentioning simple apps and know this topic better than I do. But here are my 2 cents: Sending encrypted token is safe as it is highly unlikely that user encryption is going to be broken or secret is going to be exposed. But I can imagine someone will still don't like it, for sending only db pointer is safer. And I think should we store tokens in cluster cache like REDIS, it's going to be fast and cloud native as well (e.g. in OpenShift you have REDIS operator that is easy peasy to setup), e.g. using Quarkus Redis Client, and it would be suitable for high throughput flows. Anyway, I get there must be users demand first.

I think we should close this thread. Thanks for input.

michalvavrik replied:
Done.

sberyozkin replied:
Thanks @michalvavrik We can certainly review this area of dealing with the expired tokens once users start asking

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerRunTimeConfig.java:
```diff
@@ -0,0 +1,33 @@
+     * How often should Quarkus check for expired tokens. By default, the Quarkus will check every 20 seconds.
+     */
+    @WithDefault("20")
+    Duration deleteExpiredDelay();
```
I think this is a very frequent scan period. This is how the expired token refresh works. Lets say the ID token lifespan is 1 hour, and the users want the expired ID token refreshed. If we set the cookie max age to 1 hour, the same as that of the ID token (a small leeway is added to the cookie max-age by default) then in 1 hour and 5 mins the cookie itself will be gone and no refresh will happen. 
To deal with such cases, users extend the cookie max age with the session age property lets say by 24 hours, so even if the ID token has expired - it will still reach Quarkus with the session cookie and now Quarkus can give it a try and refresh it when it sees ID token has expired. The other strategy is a pre-emptive auto-refresh - SPA in the background thread keeps pinging Quarkus every 10 mins for ex, the session cookie with ID token is sent along the way by the browser, Quarkus checks, will ID token expire within the configured period of time, say 30 mins, if yes, refresh it even if the token is still valid.

None of these strategies will work if the record with the expired ID token has gone.

I believe it should be set to say 8 hours and users can tune, we could've checked the session age extension property but this session age extension can vary across multiple tenants I guess, but users will know the average value and  set this property accordingly.

michalvavrik replied:
Done.

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerRunTimeConfig.java:
```diff
@@ -0,0 +1,33 @@
+     * Whether Quarkus should attempt to create database table where the token state is going to be stored.
+     */
+    @WithDefault("true")
+    boolean createDatabaseTableIfNotExists();
```
Is it something that is already possible to configure at the Hibernate level ?

michalvavrik replied:
Certainly, but this extension intentionally doesn't rely on Hibernate until Reactive and ORM work together. IMO we can revise it in the future.

sberyozkin replied:
OK, let me ask another question, when would users choose not to create the table if it already does not exist ?

michalvavrik replied:
My thoughts for showing how would Hibernate entity look for DB token state table was that it establishes contract - shows exactly what columns are needed.

> OK, let me ask another question, when would users choose not to create the table if it already does not exist ?

It's similar to strategy Hibernate schema strategy drop-and-create, you wouldn't use it in a production. Here, nothing bad will happen when we attempt to create table, so no harm. <======== to this point I could agree with you but:

FULL DISCLOSURE: this is my opinion from practice, I have zero statistics to support my belief

Usually you don't want to configure datasource with user that has rights for DDL operations like creating database tables. Such operations are done by DBA or someone with different credentials than are credentials provided to application. Application will usually only need DML (select ,...). So what I saw was that you have scripts to create database etc. that are executed prior application starts. And in such cases, you definitely want to create table yourself in a production.

sberyozkin replied:
@michalvavrik Makes sense

sberyozkin commented:
HI @emmanuelbernard @michalvavrik, indeed, update to the https://quarkus.io/guides/security-oidc-code-flow-authentication#handling-and-controlling-the-lifetime-of-authentication, should be part of the PR. 

Michal, I believe we just need to add another subsection there, probably named `Stateful token management`, mention that users can register custom `TokenStateManager` beans to save the tokens in DB/distributed cache etc and  Quarkus offers this extension which is activated by simply adding it as a dependency alongside the preferred reactive DB client, and list which ones are supported and show some example config. If you'd like I can contribute, thanks

michalvavrik commented:
> Michal, I believe we just need to add another subsection there, probably named `Stateful token management`, mention that users can register custom `TokenStateManager` beans to save the tokens in DB/distributed cache etc and Quarkus offers this extension which is activated by simply adding it as a dependency alongside the preferred reactive DB client, and list which ones are supported and show some example config. If you'd like I can contribute, thanks

- added config properties example
- added dependencies example for PG
- listed supported reactive sql clients (that's more for future, we support all of them, also we have build time validation, but yeah, added list)
- renamed section that I previously added to the linked guide

@sberyozkin I think it will be the easiest if you just commit & push docs change instead of suggesting line by line. I've added all I can think of and I'll be very happy if you contribute. But whatever works for you best. Thanks

sberyozkin commented on docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:
```diff
@@ -545,7 +545,8 @@ In such cases, use the `quarkus.oidc.token-state-manager.strategy` property to c
 This can occur when the ID, access, and refresh tokens are JWT tokens and the selected strategy is `keep-all-tokens` or with ID and refresh tokens when the strategy is `id-refresh-token`.
-To workaround this issue,  you can set `quarkus.oidc.token-state-manager.split-tokens=true` to create a unique session token for each token.
+To workaround this issue, you can set `quarkus.oidc.token-state-manager.split-tokens=true` to create a unique session token for each token.
+Another way to workaround this issue is to store tokens in a database. For more information, please see the <<db-token-state-manager,Stateful token management>> section of this guide.
```
```suggestion
Another way to workaround this issue is to store tokens in a database. For more information, please see the <<db-token-state-manager>> section of this guide.
```

Michelle explained it is no longer necessary to accompany such links with the text

michalvavrik replied:
Interesting, thanks.

michalvavrik replied:
Maybe you can drop it when pushing docs @sberyozkin, please? I don't want to trigger full CI just for dropping 3 words.

michalvavrik replied:
Done.

sberyozkin commented:
@michalvavrik Docs are great IMHO, and they show how well though out your PR is, this extension can align with Hibernate, which is great. I think we need to talk more about dealing with the expired tokens and explain it.

I can push some text attempting to clarify some pros and cons of the existent and the stateful approach, to avoid offering both options as equal alternatives.

Thanks

michalvavrik commented:
> I can push some text attempting to clarify some pros and cons of the existent and the stateful approach, to avoid offering both options as equal alternatives.

Thanks, please go ahead when the time is right for you.

sberyozkin commented:
@michalvavrik Hi, added a commit, let me know what you think, it can be squashed once we finalize these doc updates, thanks

michalvavrik commented on docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:
```diff
@@ -574,10 +576,16 @@ Otherwise, a random key is generated, which can be problematic if the Quarkus ap
 For example, you may want to keep the tokens in a cache cluster and have only a key stored in a session cookie.
 Note that this approach might introduce some challenges if you need to make the tokens available across multiple microservices nodes.
 
+See <<db-token-state-manager>> for the information about the default `TokenStateManager` storing the tokens in the encrypted session cookie.
```
This sentence is incorrect, you wanted to point to the default TSM

michalvavrik commented:
Hey @sberyozkin I think docs look much better with your changes, thank you. I found one incorrect link. I think it would be nice to have it clear that you contributed this docs parts as I like what you have done, but if you prefer to squash commits, whatever...

@emmanuelbernard now would be a right time for your feedback

sberyozkin commented:
@michalvavrik It is 100% your effort, I just poked a bit over the docs, let it be squashed :-). Though please wait for more feedback, indeed from @emmanuelbernard but also I'd like to ask @mkouba as he was involved into Hibernate reactive work recently, to have a look at the DB related code.

Enjoy the weekend

michalvavrik commented on extensions/oidc-db-token-state-manager/deployment/src/test/java/io/quarkus/oidc/db/token/state/manager/AbstractDbTokenStateManagerTest.java:
```diff
@@ -0,0 +1,105 @@
+
+            assertEquals("alice", textPage.getContent());
+
+            assertTokenStateCount(1);
```
@sberyozkin I can add something like that, but I can't imagine how could it happen, you see:

- default token state manager has lesser priority, so it's not used
- we assert that token state is in the database (before =1 after =0) (that's exactly the `assertTokenStateCount` you are commenting on), and it simply can't be that 2 token state managers are used due to how CDI works

Anyway, if you are convinced it is needed, I'll do that.

sberyozkin replied:
@michalvavrik 

>we assert that token state is in the database (before =1 after =0) (that's exactly the assertTokenStateCount you are commenting on), and it simply can't be that 2 token state managers are used due to how CDI works

Oh yes, sure, that must be sufficient, thanks

sberyozkin reviewed (approved):
Thanks for adding this new feature @michalvavrik.

Lets wait for some more feedback next week, cheers

sberyozkin commented:
@michalvavrik I forgot something. So, the default token state manager configuration is tenant specific, for example, lest say we have Azure and Keycloak, and for the Azure tenant we may have to split the session cookie, for Keycloak - not, etc.

This extension has a global impact on all tenants, I think this is totally fine, supporting DB storage of tokens for 1 tenant but keeping them in the session cookie for another one makes no sense.

I'll need to update the docs to mention that when this or other custom token state manager is registered, the `quarkus.oidc.token-state-manager.*` configuration will be ignored. Trying to fail the start up would be tricky as some tenants can be created dynamically. I think a note in the docs would do.

The other thing is, the test coverage is very good, but we don't test it native at all. We can try to update some of the existing integrating tests, lets say https://github.com/quarkusio/quarkus/blob/main/integration-tests/smallrye-jwt-oidc-webapp/src/test/java/io/quarkus/it/keycloak/SmallRyeJwtOidcWebAppTest.java#L65, this test suite has a mix of tests but only one of them is code flow based, so we can add this extension dependency there, instead of introducing a new integration tests module

Thanks

michalvavrik commented:
hey @sberyozkin 

> I'll need to update the docs to mention that when this or other custom token state manager is registered, the `quarkus.oidc.token-state-manager.*` configuration will be ignored. Trying to fail the start up would be tricky as some tenants can be created dynamically. I think a note in the docs would do.
> 

Agreed and good catch. I think I originally didn't ignore this config for I was validating session cookies is encrypted (`encryption-required`, `encryption-secret`), but that changed after your review and I didn't realize it, sorry. As for the other 2 `quarkus.oidc.token-state-manager` properties - they explicitly mention `Default TokenStateManager`, therefore I think it's common sense they are going to be ignored.

This could be solved by simply referring to the `Default TokenStateManager` in `strategy` and `split-tokens` as well.

> The other thing is, the test coverage is very good, but we don't test it native at all. We can try to update some of the existing integrating tests, lets say https://github.com/quarkusio/quarkus/blob/main/integration-tests/smallrye-jwt-oidc-webapp/src/test/java/io/quarkus/it/keycloak/SmallRyeJwtOidcWebAppTest.java#L65, this test suite has a mix of tests but only one of them is code flow based, so we can add this extension dependency there, instead of introducing a new integration tests module

Yeah, I didn't want to introduced new integration test module. I'll add tests to `smallrye-jwt-oidc-webapp`, thanks for the suggestion.

sberyozkin commented:
@michalvavrik Only tweaking a bit that single integration code flow test to check the count before and after logout as you do in the deployment tests should be enough, you already test it extensively in the deployment.

sberyozkin reviewed (approved):
@michalvavrik It looks perfect. Thanks for updating JavaDocs for DefaultTokenStateManager too along the way :+1: , thanks

sberyozkin commented:
Hi @yrodiere, can you please check a few of DB related classes in this PR, overall, the idea is to let uses persist 3 OIDC tokens (ID, access, refresh) in DB, query them, and remove, on demand and with a scheduled removal of expired entries.

The code in the `oidc-db-token-state-manager` extension...

Thanks

sberyozkin commented on extensions/oidc-db-token-state-manager/deployment/src/main/java/io/quarkus/oidc/db/token/state/manager/OidcDbTokenStateManagerBuildTimeConfig.java:
```diff
@@ -0,0 +1,17 @@
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+
+@ConfigMapping(prefix = "quarkus.oidc-db-token-state-manager")
```
Michal, can we try `quarkus.oidc.db-token-state-manager` - it should align better with the rest of the OIDC configuration

michalvavrik replied:
yes, I originally had it as well, but wasn't sure if conventions allow it (e.g. quarkus.oidc.* should belong to oidc extension). I agree.

I'll wait for a day if there comes another review regarding db stuff and change it. Thanks

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerRunTimeConfig.java:
```diff
@@ -0,0 +1,26 @@
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+
+@ConfigMapping(prefix = "quarkus.oidc-db-token-state-manager")
```
Also `quarkus.oidc.db-token-state-manager`

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManager.java:
```diff
@@ -0,0 +1,150 @@
+            // illegal state that shouldn't happen
+            throw new AuthenticationFailedException(SESSION_MAX_AGE_PARAM + " parameter is missing");
+        }
+        maxAge += oidcConfig.token.lifespanGrace.orElse(0);
```
`sesson_max_age` property already includes lifespan and session age extension, see https://github.com/quarkusio/quarkus/blob/main/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java#L940C34-L940C34

michalvavrik replied:
Will fix, thank you

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerInitializer.java:
```diff
@@ -0,0 +1,145 @@
+
+    private static void periodicallyDeleteExpiredTokens(Vertx vertx, Pool pool, long delayBetweenChecks) {
+        vertx
+                .setPeriodic(5000, delayBetweenChecks, new Handler<Long>() {
```
Should the first and second parameters be equal ?

michalvavrik replied:
the first parameter says before first periodic is started, wait for 5 seconds. Logic in there is that if we do many things right before app started, it can be a lot <= that's very vague intentionally, we can't know what are all the things that user will do on app start (or what extension is using), giving some times decrease overload in my mind.

> Should the first and second parameters be equal ?

It would mean that first delete is after 8 hours. What if app restarted in the middle of interval? It would mean 4 h + 8 h. But yeah, I see were you are heading to. No idea, do you prefer it to be same?

sberyozkin replied:
@michalvavrik I see, sorry

sberyozkin commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerInitializer.java:
```diff
@@ -0,0 +1,145 @@
+                    public void handle(Long aLong) {
+                        if (deleteInProgress.compareAndSet(false, true)) {
+
+                            final long deleteExpiresIn = now() - EXPIRED_EXTRA_GRACE;
```
I'm not sure it works exactly as expected. The session_max_age which was saved in the record is a relative value not related to the epoch (ex, base max age there is a difference between ID token `issued at` and `expires in` claim values). So this DELETE check will be true in most cases. 
I believe we should save `now() + session_max_age`  for this DELETE check to work ?

michalvavrik replied:
Session max age is `Indicates the number of seconds until the cookie expires.`, in the database it is stored as `expires_in=maxAgeInSeconds + epochInSeconds` (meta code). Therefore delete check is correct in my eyes, for this way, I can run DELETE statement without handling differences of time data types between databases. 

And now, I added `EXPIRED_EXTRA_GRACE` constant based on this comment on yours _The expiry should probably be calculated the same way the q_session coookie lifespan is calculated plus add a few mins to it just in case._ https://github.com/quarkusio/quarkus/issues/13239#issuecomment-932938148

sberyozkin replied:
@michalvavrik Oh I missed you were saving it as ` now() + maxAge` sure, make sense

yrodiere commented:
> Hi @yrodiere, can you please check a few of DB related classes in this PR, overall, the idea is to let uses persist 3 OIDC tokens (ID, access, refresh) in DB, query them, and remove, on demand and with a scheduled removal of expired entries.
> 
> The code in the `oidc-db-token-state-manager` extension...
> 
> Thanks

Hey, this seems to use the reactive SQL clients, and unfortunately I know as little as you do in that area, probably less :)

It's probably better to ask @tsegismont :)

sberyozkin commented:
Thanks for the feedback @yrodiere and also for giving your support to the @michalvavrik's design decision :-).

Overall, the DB related code is very simple, even I can understand it :-), I'd just like to have an expert to have a quick look, @tsegismont, if you can find a few mins to have a look it would be great, should not take much time, thanks

michalvavrik committed:
Introduce OIDC Databse Token State Manager

sberyozkin commented:
@michalvavrik I think I'm going to merge it soon enough, as I said the DB code is straightforward, a basic entity is saved/retrieved.  But we'll see what users will report and tune whatever will be required - I guess it is hard to spot any problem until users actually start stressing it. Overall it is comprehensively tested

michalvavrik commented:
> @michalvavrik I think I'm going to merge it soon enough, as I said the DB code is straightforward, a basic entity is saved/retrieved. But we'll see what users will report and tune whatever will be required - I guess it is hard to spot any problem until users actually start stressing it. Overall it is comprehensively tested

Thanks @sberyozkin . This morning I was actually looking for additional reviewers, but I didn't see between contributor to this reactive client anyone more fit than people you already asked . I can imagine some optimalization there could be, but it should be safe code anyway (also we have tests).

sberyozkin commented:
This doc build failure was not here the previous time and I've seen it happening sometimes

tsegismont commented on extensions/oidc-db-token-state-manager/runtime/src/main/java/io/quarkus/oidc/db/token/state/manager/runtime/OidcDbTokenStateManagerInitializer.java:
```diff
@@ -0,0 +1,145 @@
+
+    private static void periodicallyDeleteExpiredTokens(Vertx vertx, Pool pool, long delayBetweenChecks) {
+        vertx
+                .setPeriodic(5000, delayBetweenChecks, new Handler<Long>() {
```
`setPeriodic` returns an id that you could store in a field.

Then, in a method that observes the `ShutdownEvent`, invoke `vertx.cancelTimer(id)`.

sberyozkin replied:
Hey @tsegismont thanks for the comments, apologies I did not wait enough. As far as I recall, the timer is started on the Vert.x underlying HTTP router and therefore it will be cancelled as part of its closing sequence. I have a few of them in the OIDC code. 
Michal, if it proves to be a problem then we can handle all of them as it won't be a problem related to your PR only, please open an issue to investigate. I have another timer in the UserInfo/TokenIntrospection cache, so with this one - 3 in total, though as far as I recall those 2 I created they are activated optionally

sberyozkin replied:
>I use injected Vertx instance, therefore I thought when Vert.X is closed on shutdown, the periodic is going to be closed as well. 

I thought the same, I might've heard from someone else though.

sberyozkin replied:
@tsegismont If you'd like we can reopen this PR or indeed follow with the necessary cleanup as Michal proposed if necessary, thanks

michalvavrik replied:
@sberyozkin I suggest to wait things are cleared out before doing anything. Also I think it will be much clearer to handle it in a separate PR for changes will be easier to review.

tsegismont replied:
> hey @tsegismont , thanks for taking your time, I'll come with PR that address your comment.
> 
> Truth be told, I had this in one of my iterations, but I had 2 reasons to drop it:
> 
>     * I use injected Vertx instance, therefore I thought when Vert.X is closed on shutdown, the periodic is going to be closed as well. When I look into implementation, I don't see it though.

When a Vert.x instance is closed, all timers are indirectly stopped by calling `EventExecutorGroup#shutdownGracefully`:

https://github.com/eclipse-vertx/vert.x/blob/23eb20ffba301d735b227c604a67e5c7ec1de234/src/main/java/io/vertx/core/impl/VertxImpl.java#L839

> 
>     * I saw Sergey is already using it without cancelling in `io.quarkus.oidc.runtime.BackChannelLogoutTokenCache`, so I thought it's fine
> 
> 
> If I get it right, periodic is scheduled on `EventLoop` and if we don't cancel it, we could risk resource leaks, e.g. in DEV mode?

If you're absolutely sure the corresponding Vert.x instance is closed when this bean instance is destroyed, then you may omit to properly cancel the timer.

From my perspective it would be safer (and does not cost much) to do it explicitly.

tsegismont replied:
> Hey @tsegismont thanks for the comments, apologies I did not wait enough

No worries, really.



> @sberyozkin I suggest to wait things are cleared out before doing anything. Also I think it will be much clearer to handle it in a separate PR for changes will be easier to review.

:+1: 

I don't believe it's necessary to revert the change and reopen the PR. I hope my comment above helps and don't hesitate to ping me if it doesn't.

tsegismont reviewed (commented):
@michalvavrik @sberyozkin sorry for the late reply

The `OidcDbTokenStateManagerInitializer` looks good to me. Perhaps it can be simplified a bit by not using mutiny (not useful here), but it's only a matter of readability.

michalvavrik commented:
> If you're absolutely sure the corresponding Vert.x instance is closed when this bean instance is destroyed, then you may omit to properly cancel the timer.
> From my perspective it would be safer (and does not cost much) to do it explicitly.

I agree, fix is already merged here https://github.com/quarkusio/quarkus/pull/36366.

michalvavrik commented:
>Perhaps it can be simplified a bit by not using mutiny (not useful here), but it's only a matter of readability.

I'll add it to my list and address it eventually, frankly using Mutiny was simpler to me as I'm used to it (it's pretty much omnipresent in Quarkus). Thanks for your @tsegismont .
